[SPARK-28520][SQL] WholeStageCodegen does not work property for LocalTableScanExec

sarutak opened the pull request:
Code is not generated for LocalTableScanExec although proper situations.

If a LocalTableScanExec plan has the direct parent plan which supports WholeStageCodegen,
the LocalTableScanExec plan also should be within a WholeStageCodegen domain.
But code is not generated for LocalTableScanExec and InputAdapter is inserted for now.

```
val df1 = spark.createDataset(1 to 10).toDF
val df2 = spark.createDataset(1 to 10).toDF
val df3 = df1.join(df2, df1("value") === df2("value"))
df3.explain(true)

...

== Physical Plan ==
*(1) BroadcastHashJoin [value#1], [value#6], Inner, BuildRight
:- LocalTableScan [value#1]                                             // LocalTableScanExec is not within a WholeStageCodegen domain
+- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)))
   +- LocalTableScan [value#6]
```

```
scala> df3.queryExecution.executedPlan.children.head.children.head.getClass
res4: Class[_ <: org.apache.spark.sql.execution.SparkPlan] = class org.apache.spark.sql.execution.InputAdapter
```

For the current implementation of LocalTableScanExec, codegen is enabled in case `parent` is not null
but `parent` is set in `consume`, which is called after `insertInputAdapter` so it doesn't work as intended.

After applying this cnahge, we can get following plan, which means LocalTableScanExec is within a WholeStageCodegen domain.

```
== Physical Plan ==
*(1) BroadcastHashJoin [value#63], [value#68], Inner, BuildRight
:- *(1) LocalTableScan [value#63]
+- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)))
   +- LocalTableScan [value#68]

## How was this patch tested?

New test cases are added into WholeStageCodegenSuite.

sarutak committed:
Fix CollapseCodegenStages#insertWholeStageCodegen to generate code for LocalTableScanExec properly

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13305/
Test PASSed.

SparkQA commented:
**[Test build #108205 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108205/testReport)** for PR 25260 at commit [`24d51ba`](https://github.com/apache/spark/commit/24d51ba1c30472cffc4e44641ece2c4e76e54139).

SparkQA commented:
**[Test build #108205 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108205/testReport)** for PR 25260 at commit [`24d51ba`](https://github.com/apache/spark/commit/24d51ba1c30472cffc4e44641ece2c4e76e54139).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108205/
Test PASSed.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:
```diff
@@ -895,6 +895,10 @@ case class CollapseCodegenStages(
+      case plan: LocalTableScanExec =>
+        // Do not make LogicalTableScanExec the root of WholeStageCodegen
+        // to support the fast driver-local collect/take paths.
+        plan
```
cc: @juliuszsompolski

dongjoon-hyun replied:
What about `branch-2.4/2.3`, @sarutak ?

maropu replied:
IIUC LocalTableScan doesn't support codegen in these branches.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/LocalTableScanExec.scala:
```diff
@@ -80,8 +80,7 @@ case class LocalTableScanExec(
 
-  // Do not codegen when there is no parent - to support the fast driver-local collect/take paths.
-  override def supportCodegen: Boolean = (parent != null)
+  override def supportCodegen: Boolean = true
```
We don't need this cuz CodegenSupport already sets true.

sarutak replied:
Ah, exactly.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:
```diff
@@ -325,4 +326,68 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
+
+    // Ensure WholeStageCodegenExec is not inserted and
+    // LocalTableScanExec is still the root.
+    assert(!rootOfExecutedPlan.isInstanceOf[WholeStageCodegenExec],
```
I think we don't need this test and the test below is enought for this fix: 
```
    // Ensure WholeStageCodegenExec is not inserted and
    // LocalTableScanExec is still the root.
    assert(rootOfExecutedPlan.isInstanceOf[LocalTableScanExec],
      "LocalTableScanExec should be still the root.")

    // Case2: The parent of a LocalTableScanExec supports WholeStageCodegen.
```

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:
```diff
@@ -325,4 +326,68 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
+          case InputAdapter(_: LocalTableScanExec, _) => true
+          case _ => false
+        }.isEmpty,
+        "InputAdapter should not be inserted.")
```
ditto: I think we don't need the test below;
```
      // No more InputAdapter inserted between LocalTableScanExec and its parent.
      assert(
        executedPlan.find {
          case InputAdapter(_: LocalTableScanExec, _) => true
          case _ => false
        }.isEmpty,
        "InputAdapter should not be inserted.")
```

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:
```diff
@@ -325,4 +326,68 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
+        "InputAdapter should not be inserted.")
+    }
+
+    // Case3: The parent of a plan of LocalTableScanExec does not support WholeStageCodegen.
```
IMO the cases 1/2 are enough for tests.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:
```diff
@@ -325,4 +326,68 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
+    val rightDF = spark.createDataset(1 to 10).toDF
+
+    // Force BroadcastHasJoin enabled
+    withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> Long.MaxValue.toString) {
```
nit: How about using `HashAggregate` for the test? If so, we could remove this config set;
```
scala> Seq(1, 2, 3).toDF("a").groupBy().sum().explain
== Physical Plan ==
*(2) HashAggregate(keys=[], functions=[sum(cast(a#52 as bigint))])
+- Exchange SinglePartition
   +- *(1) HashAggregate(keys=[], functions=[partial_sum(cast(a#52 as bigint))])
      +- LocalTableScan [a#52]
```

sarutak replied:
Replaced with HashAggregate.

maropu reviewed (changes requested):
nice catch, LGTM except for the minor comments @cloud-fan @juliuszsompolski

juliuszsompolski commented:
Good catch. Thanks for the fix!

sarutak committed:
Remove redundant tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13351/
Test PASSed.

SparkQA commented:
**[Test build #108249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108249/testReport)** for PR 25260 at commit [`da3f193`](https://github.com/apache/spark/commit/da3f193fa3c1d58421b0e21707bc922980f42c40).

maropu commented:
cc: @dongjoon-hyun

SparkQA commented:
**[Test build #108249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108249/testReport)** for PR 25260 at commit [`da3f193`](https://github.com/apache/spark/commit/da3f193fa3c1d58421b0e21707bc922980f42c40).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108249/
Test PASSed.

dongjoon-hyun commented:
Thank you for pinging me, @maropu . I'll take a look today.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:
```diff
@@ -325,4 +326,33 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
     }
   }
+
+  test("WholeStageCodegen does not work properly for LocalTableScanExec") {
```
Could you use `SPARK-28520` prefix since this is a bug fix?

sarutak replied:
Thanks. I added the prefix.

sarutak committed:
Add "SPARK-28520" prefix to the test name

SparkQA commented:
**[Test build #108265 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108265/testReport)** for PR 25260 at commit [`050c4c4`](https://github.com/apache/spark/commit/050c4c4e7f3bd37b2815f89859023128f967e57c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13367/
Test PASSed.

SparkQA commented:
**[Test build #108265 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108265/testReport)** for PR 25260 at commit [`050c4c4`](https://github.com/apache/spark/commit/050c4c4e7f3bd37b2815f89859023128f967e57c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108265/
Test FAILed.

sarutak commented:
retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13370/
Test PASSed.

SparkQA commented:
**[Test build #108268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108268/testReport)** for PR 25260 at commit [`050c4c4`](https://github.com/apache/spark/commit/050c4c4e7f3bd37b2815f89859023128f967e57c).

SparkQA commented:
**[Test build #108268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108268/testReport)** for PR 25260 at commit [`050c4c4`](https://github.com/apache/spark/commit/050c4c4e7f3bd37b2815f89859023128f967e57c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108268/
Test PASSed.

maropu commented:
Thanks! Merged to master.

maropu commented:
Thanks, all @sarutak @juliuszsompolski @dongjoon-hyun @kiszk

rednaxelafx reviewed (commented):
Post-hoc LGTM. Thanks a lot for noticing this, Saruta-san!

There are a few things I'm wondering about, though. Is `LocalTableScanExec` the only case where we're doing this kind of "if root then don't codegen" cutoff? If there are other cases, it might make more sense to extract a trait and make the `insertWholeStageCodegen` logic cleaner and more flexible.

Also, what about the case when the immediate parent of a `LocalTableScanExec` is a Spark stage boundary, e.g. an `Exchange`? That case seems to be covered by this PR already (which doesn't insert a `WholeStageCodegenExec` right above `LocalTableScanExec`. Is that correct?

sarutak commented:
Thanks @reactormonk for the comment!

>  There are a few things I'm wondering about, though. Is LocalTableScanExec the only case where we're doing this kind of "if root then don't codegen" cutoff? If there are other cases, it might make more sense to extract a trait and make the insertWholeStageCodegen logic cleaner and more flexible.

Yes for now. First I considered adding new trait but I found the case we need to cut off was only `LocaTableScanExec` so I decided to change `insertWholeStageCodegen` like a `SMJ` case in `insertInputAdapter`.  But it's  worth re-considering to add new trait. I'll try it.

> Also, what about the case when the immediate parent of a LocalTableScanExec is a Spark stage boundary, e.g. an Exchange? That case seems to be covered by this PR already (which doesn't insert a WholeStageCodegenExec right above LocalTableScanExec. Is that correct?

Yes, this PR cover the case you mentioned. Actually, I added a test case for test that case but I removed in response to the review comment. Do you think it's still good to add a test case for the case?

sarutak commented:
Sorry, I withdraw my comment above. 
Yes, this PR covers the case which `LocalTableScanExec` has parent of `Exchange` but there is no test case for testing it. Actually I added such a test case but removed in response to comment. Do you think it's good to add test case for that case? If so, I'll add it.
